fmt/src/git.rs

```diff
@@ -34,38 +38,45 @@ impl GitHelper {
             return Ok(None);
         }
 
-        match Repository::discover(basedir) {
+        match gix::open(basedir) {
             Ok(repo) => {
-                if repo.workdir().is_none() {
+                if repo.worktree().is_none() {
+                    let message = "bare repository detected";
                     if config.ignore.is_auto() {
-                        info!("git.ignore=auto is resolved to disabled; bare repository detected");
+                        info!("git.ignore=auto is resolved to fallback; {message}");
                         Ok(None)
                     } else {
-                        InvalidConfigSnafu {
-                            message: "Git repository is bare".to_string(),
-                        }
-                        .fail()
+                        InvalidConfigSnafu { message }.fail()
                     }
                 } else {
                     info!("git.ignore=auto is resolved to enabled");
-                    Ok(Some(GitHelper { repo }))
+                    let state = State::new(Kind::Sha1);
+                    Ok(Some(GitHelper { repo, state }))
                 }
             }
             Err(err) => {
                 if config.ignore.is_auto() {
                     info!(?err, "git.ignore=auto is resolved to disabled");
                     Ok(None)
                 } else {
-                    Err(err).context(GitOpSnafu)
+                    Err(GixOpenOpSnafu {}.into_error(Box::new(err)))
                 }
             }
         }
     }
 
-    pub fn ignored(&self, path: &Path) -> Result<bool> {
+    pub fn ignored(&self, path: &Path, is_dir: bool) -> Result<bool> {
         let path = path.canonicalize().context(ResolveAbsolutePathSnafu {
             path: path.display().to_string(),
         })?;
-        self.repo.is_path_ignored(path).context(GitOpSnafu)
+        let mut stack = self
+            .repo
+            .excludes(&self.state, None, Source::default())
```

Comment:
To not have to deal with the index, I suggested `repo.worktree()?.excludes()` previously. Otherwise, the index can be opened with `repo.index()?`.

However, doing this is no better, performance-wise, than `git2` as it will have to do a lot of repetitive work each time the path is queried.

It's probably OK to keep it at first just to get it to work.

Reply:
Yes. I found that the lifecycle mark make it hard to hold the AttributeStack and I'm trying to get rid of git2 first.

If we can make it work as better as current, we can make further improvements.